### PR TITLE
chore: mod path of image pull secret

### PR DESCRIPTION
**Pull request by chriswldenyer**
Move the image pull secret to be under image, not base.

## Description

Move the image pull secret to be under image, not base.
This is the more logical location.

## Regression
No

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.

### CODE REVIEW

File: chart/templates/_helpers.tpl [hunk 1]
```diff
@@ -77,8 +77,16 @@ Usage:
...
 {{- define "base_pull_secrets" -}}
-    {{- if .Values.base.imagePullSecrets.enabled }}
-    {{- include "render" (dict "value" .Values.base.imagePullSecrets.secrets "context" $) | nindent 8 }}
+    {{- if .Values.base.imagePullSecrets }}
+        {{- if .Values.base.imagePullSecrets.enabled }}
+        {{- include "render" (dict "value" .Values.base.imagePullSecrets.secrets "context" $) | nindent 8 }}
+        {{- end }}
+    {{- else -}}
+        {{- if .Values.image.pullSecrets }}
+            {{- if .Values.image.pullSecrets.enabled }}
```
Comment (tiagolobocastro) [inline chart/templates/_helpers.tpl]: I think this would work?
```suggestion
        {{- if (.Values.image.pullSecrets).enabled }}
```

**Review by tiagolobocastro — APPROVED** [review]
Please run helm upgrade manually (previous install with secrets enabled) to ensure it's working

File: chart/values.yaml [hunk 2]
```diff
@@ -26,6 +26,14 @@ image:
   # -- ImagePullPolicy for our images
   pullPolicy: Always
 
+  # docker-secrets required to pull images if the container registry from image.Registry is protected
+  pullSecrets:
+    # -- Enable pullSecrets for pulling our container images
+    enabled: false
+    # -- Name of the pullSecret in the installed namespace
+    secrets:
+      - name: login
```
Comment (dsharma-dc) [inline chart/values.yaml]: This (secrets)name is not going to change in this repo I'd believe?

Comment (tiagolobocastro) [inline chart/values.yaml]: Looking at bitnami, they simply pass an array of secrets, presumably meaning its existence implies the enabled to true.
So I suggest we simply do the same:

```suggestion
  pullSecrets: []
```

**Comment by chriswldenyer** [comment]
bors try

File: chart/templates/_helpers.tpl (hunk 1 above)
Comment (chriswldenyer) [reply chart/templates/_helpers.tpl]: Helm doesn't seem to like that shortcut:
helm template mayastor .
Error: template: mayastor/templates/mayastor/agents/ha/ha-node-daemonset.yaml:33:12: executing "mayastor/templates/mayastor/agents/ha/ha-node-daemonset.yaml" at <include "base_pull_secrets" .>: error calling include: template: mayastor/templates/_helpers.tpl:80:18: executing "base_pull_secrets" at <.Values.base.imagePullSecrets.enabled>: nil pointer evaluating interface {}.enabled

File: chart/values.yaml (hunk 2 above)
Comment (chriswldenyer) [reply chart/values.yaml]: fixed

File: chart/values.yaml [hunk 3]
```diff
@@ -26,6 +26,9 @@ image:
   # -- ImagePullPolicy for our images
   pullPolicy: Always
 
+  # docker-secrets required to pull images if the container registry from image.Registry is protected
```
Comment (tiagolobocastro) [inline chart/values.yaml]:
```suggestion
  # -- docker-secrets required to pull images if the container registry from image.registry is protected
```

**Review by tiagolobocastro — APPROVED** [review]

File: chart/values.yaml (hunk 2 above)
Comment (chriswldenyer) [reply chart/values.yaml]: Actually there is no default name any more in the latest change.

File: chart/values.yaml (hunk 3 above)
Comment (chriswldenyer) [reply chart/values.yaml]: fixed

**Comment by chriswldenyer** [comment]
bors try

**Review by Abhinandan-Purkait — APPROVED** [review]

File: chart/templates/_helpers.tpl [hunk 4]
```diff
@@ -77,8 +77,12 @@ Usage:
...
 */}}
 {{- define "base_pull_secrets" -}}
-    {{- if .Values.base.imagePullSecrets.enabled }}
-    {{- include "render" (dict "value" .Values.base.imagePullSecrets.secrets "context" $) | nindent 8 }}
+    {{- if .Values.base.imagePullSecrets }}
+        {{- if .Values.base.imagePullSecrets.enabled }}
+        {{- include "render" (dict "value" .Values.base.imagePullSecrets.secrets "context" $) | nindent 8 }}
+        {{- end }}
+    {{- else -}}
+    {{- include "render" (dict "value" .Values.image.pullSecrets "context" $) | nindent 8 }}
```
Comment (tiagolobocastro) [inline chart/templates/_helpers.tpl]: As discussed, this needs changing

Comment (chriswldenyer) [reply chart/templates/_helpers.tpl]: Updated, as suggested.

**Review by niladrih — APPROVED** [review]
Configuration lives through helm upgrade, no upgrade changes required.

**Comment by chriswldenyer** [comment]
Re-requested reviews because of a change in the logic of the pull-secrets helper function.

**Review by tiagolobocastro — APPROVED** [review]

**Review by dsharma-dc — APPROVED** [review]

**Comment by chriswldenyer** [comment]
One further change - use default secret name of "login" if the legacy method is used to enable the secret and the name is not specified.

File: chart/templates/_helpers.tpl [hunk 5]
```diff
@@ -77,8 +77,20 @@ Usage:
...
-    {{- include "render" (dict "value" .Values.base.imagePullSecrets.secrets "context" $) | nindent 8 }}
+    {{- if (not (empty .Values.image.pullSecrets)) }}
+        {{- range .Values.image.pullSecrets | uniq -}}
+            {{ nindent 8 "- name:" }} {{ . }}
+        {{- end }}
+    {{- else -}}
+        {{- if .Values.base.imagePullSecrets }}
+            {{- if .Values.base.imagePullSecrets.enabled }}
+                {{- if (empty .Values.base.imagePullSecrets.secrets) }}
+                    {{ nindent 8 "- name: login" }}
```
Comment (niladrih) [inline chart/templates/_helpers.tpl:88]: Is this really necessary? If neither the new nor the old configuration keys are being used, should we assume configuration for a user? This feels like it is a good candidate for a failure case.

**Review by niladrih — COMMENTED** [review]
Left a comment.

File: chart/templates/_helpers.tpl (hunk 5 above)
Comment (chriswldenyer) [reply chart/templates/_helpers.tpl:88]: I was thinking of the case where the user is specifying --set base.imagePullSecrets=true but relying of the default name of "login", and having defined a secret called login in the cluster. In which case the name would no longer be defined in the updated values.yaml file.
I couldn't say how likely this scenario would be for users, but is possible AFAICT.

Comment (niladrih) [reply chart/templates/_helpers.tpl:88]: Okay, yeah that case makes sense to me. Good idea to leave this in I guess.

**Review by niladrih — APPROVED** [review]

**Review by Abhinandan-Purkait — APPROVED** [review]

**Review by tiagolobocastro — APPROVED** [review]

**Comment by chriswldenyer** [comment]
bors merge

**Pull request merged**